{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 5: Proteins\n",
    "\n",
    "## Python Dictionaries: Translate DNA/RNA to AA\n",
    "\n",
    "Write a Python program called `translate.py` that translates a given DNA/RNA sequence to amino acids.\n",
    "The input sequence should be a positional argument.\n",
    "The codon table for the `-c` or `--codons` option is _required_ and should be a valid, readable file.\n",
    "The `-o` or `--output` option (default `'out.txt'`) will name a file to write the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this cell to make sure this assignment is up to date\n",
    "%cd ~/be434-Spring2024\n",
    "!git pull --no-edit upstream main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting Started with new.py\n",
    "\n",
    "Let's start out by using new.py to create a program template for us.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the `translate.py` using `new.py`\n",
    "%cd ~/be434-Spring2024/assignments/05_proteins\n",
    "!new.py -p 'Translate DNA/RNA to AA' translate.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should see the following:\n",
    "\n",
    "```\n",
    "$ new.py -p 'Translate DNA/RNA to AA' translate.py\n",
    "Done, see new script \"translate.py.\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Instructions\n",
    "\n",
    "### Program Overview\n",
    "\n",
    "When run with no inputs, the program should print a brief usage:\n",
    "\n",
    "```\n",
    "$ ./translate.py\n",
    "usage: translate.py [-h] -c FILE [-o FILE] str\n",
    "translate.py: error: the following arguments are required: str, -c/--codons\n",
    "```\n",
    "\n",
    "Or a longer usage for the `-h` or `--help` flags:\n",
    "\n",
    "```\n",
    "$ ./translate.py -h\n",
    "usage: translate.py [-h] -c FILE [-o FILE] str\n",
    "\n",
    "Translate DNA/RNA to proteins\n",
    "\n",
    "positional arguments:\n",
    "  str                   DNA/RNA sequence\n",
    "\n",
    "optional arguments:\n",
    "  -h, --help            show this help message and exit\n",
    "  -c FILE, --codons FILE\n",
    "                        A file with codon translations (default: None)\n",
    "  -o FILE, --outfile FILE\n",
    "                        Output filename (default: out.txt)\n",
    "\n",
    "```\n",
    "\n",
    "### Output\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "Output written to \"out.txt\".\n",
    "$ cat out.txt\n",
    "ELHRSPG\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Detailed Output Specifications\n",
    "\n",
    "Part of writing good code is testing the input that a user gives you to make sure it is valid. \n",
    "You should do this for every program you write to ensure that your code is robust and no unexpected errors occur.\n",
    "As part of this assignment, we will be testing the arguments.\n",
    "\n",
    "The program should reject a `--codons` argument that fails to name a file:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c lkdflk ACTG\n",
    "usage: translate.py [-h] -c FILE [-o FILE] str\n",
    "translate.py: error: argument -c/--codons: can't open 'lkdflk': \n",
    "[Errno 2] No such file or directory: 'lkdflk'\n",
    "```\n",
    "\n",
    "If given good input, write the results to the proper output file.\n",
    "Note here that the input string is lowercase:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "Output written to \"out.txt\".\n",
    "```\n",
    "\n",
    "And now the `out.txt` file should contain the following:\n",
    "\n",
    "```\n",
    "$ cat out.txt\n",
    "ELHRSPG\n",
    "```\n",
    "\n",
    "Note that the `-o` flag should make the program write the output to a different file.\n",
    "Note here that the input sequence is uppercase:\n",
    "\n",
    "```\n",
    "$ ./translate.py UGGCCAUGGCGCCCAGAACUG --codons inputs/codons.rna -o foo.txt\n",
    "Output written to \"foo.txt\".\n",
    "```\n",
    "\n",
    "Your program will be given the wrong codon table for a given sequence type, e.g., a DNA table for an RNA sequence:\n",
    "\n",
    "```\n",
    "$ ./translate.py UGGCCAUGGCGCCCAGAACUG --codons codons.dna\n",
    "Output written to \"out.txt\".\n",
    "```\n",
    "\n",
    "This means that your program will be unable to find some of the codons in which case it should print a `-` for a missing codon:\n",
    "\n",
    "```\n",
    "$ cat out.txt\n",
    "-P-RPE-\n",
    "```\n",
    "\n",
    "If you are creating a dictionary from the codon table, e.g.:\n",
    "\n",
    "```\n",
    "$ head -3 codons.rna\n",
    "AAA\tK\n",
    "AAC\tN\n",
    "AAG\tK\n",
    "```\n",
    "\n",
    "Such that you have something like this:\n",
    "\n",
    "```\n",
    ">>> codons = dict(AAA='K', AAC='N', AAG='K')\n",
    "```\n",
    "\n",
    "Everything is fine as long as you ask for codons that are defined but will fail at runtime if you ask for a codon that does not exist:\n",
    "\n",
    "```\n",
    ">>> codons['AAC']\n",
    "'N'\n",
    ">>> codons['AAT']\n",
    "Traceback (most recent call last):\n",
    "  File \"<stdin>\", line 1, in <module>\n",
    "KeyError: 'AAT'\n",
    "```\n",
    "\n",
    "What dictionary access method will prevent you from creating an exception if you ask for a key that doesn't exist and will also allow you to set a default value?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Time to write some code!\n",
    "\n",
    "Open the script here in VS Code in be434-Spring2024 -> assignments -> 05_proteins -> translate.py \n",
    "\n",
    "Write/edit the code using the instructions above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hint 1: Getting Started\n",
    "\n",
    "You should have used `new.py` above to create the program:\n",
    "\n",
    "```\n",
    "$ new.py -p 'Translate DNA/RNA' translate.py\n",
    "Done, see new script \"translate.py\".\n",
    "```\n",
    "\n",
    "Modify the `get_args()` function to have the following arguments:\n",
    "\n",
    "1. A positional `str` argument which is the DNA/RNA sequence to translate\n",
    "2. A `-c|--codons` argument with `type=argparse.FileType('rt')`. \n",
    "Normally, options are _optional_, but in this case, please use `required=True` to force the user to provide a value.\n",
    "3. An `-o|--outfile` argument with `type=argparse.FileType('wt')` to make it a _writable text_ file. \n",
    "Use `default='out.txt'` so that this is optional.\n",
    "\n",
    "Update `main()` to print the values:\n",
    "\n",
    "```\n",
    "def main():\n",
    "    args = get_args()\n",
    "    print('seq =', args.sequence)\n",
    "    print('codons =', args.codons)\n",
    "    print('outfile =', args.outfile)\n",
    "```\n",
    "\n",
    "Run your program with valid input and verify you see the following:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "seq = gaactacaccgttctcctggt\n",
    "codons = <_io.TextIOWrapper name='inputs/codons.dna' mode='rt' encoding='UTF-8'>\n",
    "outfile = <_io.TextIOWrapper name='out.txt' mode='wt' encoding='UTF-8'>\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hint 2: Extracting k-mers\n",
    "\n",
    "A k-mer is a `k`-length sequence of contigous characters in a string (a \"mer\" like \"polymer\").\n",
    "Here is a piece of code that you can use to extract all the 3-mers from the sequence `seq`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add this code to your program to extract all the 3-mers from the sequence `seq`\n",
    "seq = 'actga'\n",
    "k = 3\n",
    "for codon in [seq[i:i + k] for i in range(0, len(seq) - k + 1)]:\n",
    "    print(codon)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When you run the code you should see something like this:\n",
    "\n",
    "```\n",
    ">>> seq = 'actga'\n",
    ">>> k = 3\n",
    ">>> for codon in [seq[i:i + k] for i in range(0, len(seq) - k + 1)]:\n",
    "...     print(codon)\n",
    "...\n",
    "act\n",
    "ctg\n",
    "tga\n",
    "```\n",
    "\n",
    "You will then use the `codon` as a key to lookup the amino acid translation from the `codons` dictionary you build using the input file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hint 3: Creating the Codon Lookup Table\n",
    "\n",
    "The next step is to create a Python dictionary that holds the translation from a codon like \"AAA\" to an amino acid like \"K\".\n",
    "This information will come from an input file such as _inputs/codons.dna_.\n",
    "In the following command, I'm using `head` to show the first five lines of the file so that you can see the codons and residues occur as space-separated pair on each line:\n",
    "\n",
    "```\n",
    "$ head -n 5 inputs/codons.dna\n",
    "AAA K\n",
    "AAC N\n",
    "AAG K\n",
    "AAT N\n",
    "ACA T\n",
    "```\n",
    "\n",
    "The `args.codons` argument will be a _filehandle_, so you don't need to use the `open()` function.\n",
    "Instead, you can read each line of the argument using a `for` loop.\n",
    "Update your `main()` to read each line from the file, strip the trailing whitespace, and use `str.split()` to break the line on whitespace:\n",
    "\n",
    "```\n",
    "def main():\n",
    "    args = get_args()\n",
    "    for line in args.codons:\n",
    "        print(line.rstrip().split())\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Verify your code to check what you have so far\n",
    "\n",
    "Run your program and verify that it will print the following output from the codons file.\n",
    "Note that I'm connecting the `STDOUT` (standard out) of this program to `STDIN` (standard in) of `head -n 5` to show only the first five lines:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt | head -n 5\n",
    "['AAA', 'K']\n",
    "['AAC', 'N']\n",
    "['AAG', 'K']\n",
    "['AAT', 'N']\n",
    "['ACA', 'T']\n",
    "```\n",
    "\n",
    "For the following code, I'm using this import from the `pprint` (pretty-print) module:\n",
    "\n",
    "```\n",
    "from pprint import pprint\n",
    "```\n",
    "\n",
    "Now consider what changes your code would need to do to create a dictionary where the keys are the codons and the values are the residues.\n",
    "\n",
    "```\n",
    "def main():\n",
    "    args = get_args()\n",
    "    codon_table = {}\n",
    "    for line in args.codons:\n",
    "        # what goes here?\n",
    "\n",
    "    pprint(codon_table) # pretty print\n",
    "```\n",
    "\n",
    "When I run my version, I see the following output, which is truncated for brevity's sake:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "{'AAA': 'K',\n",
    " 'AAC': 'N',\n",
    " 'AAG': 'K',\n",
    " 'AAT': 'N',\n",
    " 'ACA': 'T',\n",
    " ...\n",
    " 'TTC': 'F',\n",
    " 'TTG': 'L',\n",
    " 'TTT': 'F'}\n",
    "```\n",
    "\n",
    "Verify it also works with the RNA codons:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.rna gaactacaccgttctcctggt\n",
    "{'AAA': 'K',\n",
    " 'AAC': 'N',\n",
    " 'AAG': 'K',\n",
    " 'AAU': 'N',\n",
    " 'ACA': 'T',\n",
    " 'ACC': 'T',\n",
    " 'ACG': 'T',\n",
    " 'ACU': 'T',\n",
    " ...\n",
    " 'UGU': 'C',\n",
    " 'UUA': 'L',\n",
    " 'UUC': 'F',\n",
    " 'UUG': 'L',\n",
    " 'UUU': 'F'}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verify your code for DNA\n",
    "%cd ~/be434-Spring2024/assignments/05_proteins\n",
    "!./translate.py -c inputs/codons.dna gaactacaccgttctcctggt | head -n 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verify your code for RNA\n",
    "!./translate.py -c inputs/codons.rna gaactacaccgttctcctggt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hint 4 Translating the Sequence\n",
    "\n",
    "Next, you must iterate through the codons of the input sequence, which you can do with the following code:\n",
    "\n",
    "```\n",
    "def main():\n",
    "    args = get_args()\n",
    "    codon_table = {}\n",
    "    for line in args.codons:\n",
    "        # fancy bits here\n",
    "\n",
    "    k = 3\n",
    "    seq = args.sequence\n",
    "    for codon in [seq[i:i + k] for i in range(0, len(seq), k)]:\n",
    "        print(codon)\n",
    "```\n",
    "\n",
    "At this point, your program should be able to print the following:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "gaa\n",
    "cta\n",
    "cac\n",
    "cgt\n",
    "tct\n",
    "cct\n",
    "ggt\n",
    "```\n",
    "\n",
    "Alter this code to print the codon and then what the codon translates to so that it will print the following:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "GAA E\n",
    "CTA L\n",
    "CAC H\n",
    "CGT R\n",
    "TCT S\n",
    "CCT P\n",
    "GGT G\n",
    "```\n",
    "\n",
    "Once you've got that, make your program print the protein sequence on one line:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "ELHRSPG\n",
    "```\n",
    "\n",
    "Finally, make this string print into the `args.outfile` instead of to `STDOUT`, and have your program report the name of the output file:\n",
    "\n",
    "```\n",
    "$ ./translate.py -c inputs/codons.dna gaactacaccgttctcctggt\n",
    "Output written to \"out.txt\".\n",
    "$ cat out.txt\n",
    "ELHRSPG\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing\n",
    "\n",
    "As you write your code, you can test it along the way to make sure that you are passing all of the tests for the homework. \n",
    "We will use the test suite that is included with the assignment to test that you are meeting all of the requirements in the instructions above. \n",
    "You will find the steps below to test your code. Note that you can also run these commands from a \"shell\" within the VS Code GUI. Or, you can run them here... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format your code to make it beautiful (this is called linting)\n",
    "!black ~/be434-Spring2024/assignments/05_proteins/translate.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now run the tests on your code\n",
    "%cd ~/be434-Spring2024/assignments/05_proteins\n",
    "!make test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A passing test suite looks like this:\n",
    "\n",
    "```\n",
    "$ conda activate /groups/bhurwitz/bh_class/be434/be434-conda\n",
    "$ make test\n",
    "pytest -xv --pylint --disable-warnings test.py translate.py\n",
    "============================= test session starts ==============================\n",
    "...\n",
    "collected 14 items\n",
    "\n",
    "test.py::PYLINT SKIPPED (file(s) previously passed pylint checks)        [  7%]\n",
    "test.py::FLAKE8 SKIPPED (file(s) previously passed FLAKE8 checks)        [ 14%]\n",
    "test.py::test_exists PASSED                                              [ 21%]\n",
    "test.py::test_usage PASSED                                               [ 28%]\n",
    "test.py::test_no_args PASSED                                             [ 35%]\n",
    "test.py::test_missing_input PASSED                                       [ 42%]\n",
    "test.py::test_missing_codons PASSED                                      [ 50%]\n",
    "test.py::test_bad_codon_file PASSED                                      [ 57%]\n",
    "test.py::test_good_input1 PASSED                                         [ 64%]\n",
    "test.py::test_good_input2 PASSED                                         [ 71%]\n",
    "test.py::test_good_input3 PASSED                                         [ 78%]\n",
    "test.py::test_good_input4 PASSED                                         [ 85%]\n",
    "translate.py::PYLINT SKIPPED (file(s) previously passed pylint checks)   [ 92%]\n",
    "translate.py::FLAKE8 SKIPPED (file(s) previously passed FLAKE8 checks)   [100%]\n",
    "\n",
    "======================== 10 passed, 4 skipped in 0.25s =========================\n",
    "\n",
    "Your grade is whatever percentage of tests your code passes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Uploading your code to GitHub\n",
    "\n",
    "Once you have written the code for your assignment, and are passing all of the tests above, you are ready to submit the assignment for grading. \n",
    "Use the steps below to submit your code to GitHub.\n",
    "\n",
    "* Note, if you are having any issues with passing tests, and need help, you can also submit the code with a different commit message like the following. \n",
    "\n",
    "```\n",
    "git commit -m \"test_xxx failing for 05_proteins\"\n",
    "```\n",
    "\n",
    "Once you have done that, send a private slack message to me @bhurwitz to let me know you submitted code and need help.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Submit your code to Github\n",
    "%cd\n",
    "%cd be434-Spring2024\n",
    "!git add -A && git commit -m \"Submitting 05_proteins for grading\"\n",
    "!git push"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great job! You are done with this assignment.\n",
    "\n",
    "## Authors\n",
    "\n",
    "Bonnie Hurwitz <bhurwitz@arizona.edu> and Ken Youens-Clark <kyclark@gmail.com>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
